src/Cache/FileSystemCacheItemPool.php

```diff
@@ -0,0 +1,226 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Auth\Cache;
+
+use ErrorException;
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+
+class FileSystemCacheItemPool implements CacheItemPoolInterface
+{
+    /**
+     * @var string
+     */
+    private string $cachePath;
+
+    /**
+     * @var array<CacheItemInterface>
+     */
+    private array $buffer = [];
+
+    /**
+     * Creates a FileSystemCacheItemPool cache that stores values in local storage
+     *
+     * @param string $path The string representation of the path where the cache will store the serialized objects.
+     */
+    public function __construct(string $path)
+    {
+        $this->cachePath = $path;
+
+        if (is_dir($this->cachePath)) {
+            return;
+        }
+
+        if (!mkdir($this->cachePath)) {
+            throw new ErrorException("Cache folder couldn't be created.");
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItem(string $key): CacheItemInterface
+    {
+        if (!$this->validKey($key)) {
+            throw new InvalidArgumentException("The key '$key' is not valid. The key should follow the pattern |^[a-zA-Z0-9_\.! ]+$|");
+        }
+
+        $itemPath = $this->cacheFilePath($key);
+
+        if (!file_exists($itemPath)) {
+            return new TypedItem($key);
+        }
+
+        $serializedItem = file_get_contents($itemPath);
+
+        if ($serializedItem === false) {
+            return new TypedItem($key);
+        }
+
+        return unserialize($serializedItem);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return iterable<CacheItemInterface> An iterable object containing all the
+     *   A traversable collection of Cache Items keyed by the cache keys of
+     *   each item. A Cache item will be returned for each key, even if that
+     *   key is not found. However, if no keys are specified then an empty
+     *   traversable MUST be returned instead.
+     */
+    public function getItems(array $keys = []): iterable
+    {
+        $result = [];
+
+        foreach ($keys as $key) {
+            $result[$key] = $this->getItem($key);
+        }
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function save(CacheItemInterface $item): bool
+    {
+        if (!$this->validKey($item->getKey())) {
+            return false;
+        }
+
+        $itemPath = $this->cacheFilePath($item->getKey());
+        $serializedItem = serialize($item);
```

Comment:
is this how it's done in other file caching libraries? Just curious

Reply:
It depends on what the cache is using, in this case as we're using files it is necessary to serialize the object we want to cache.

In both the `fileSystem` library and symfony's `filesystemAdapter` `serialize` is used:
* Here: https://github.com/php-cache/filesystem-adapter/blob/f1faaae40aaa696ef899cef6f6888aedb90b419b/FilesystemCachePool.php#L113
* And here: https://github.com/symfony/symfony/blob/99b25a1a9de655961ac7c8fbf18e5d00f84ed798/src/Symfony/Component/Cache/Marshaller/DefaultMarshaller.php#L44

Reply:
I can see why it's important to serialize the contents of the cache item to the cache, but serializing the cache item itself is what struck me as odd. For instance, I do not think we are supposed to preserve `expiresAt` each time, but in this case it'll be preserved as part of the `Item`. 

I believe this is the case with other caching libraries - they serialize the contents of the Item and then create a new one with those contents on hit.

Reply:
See https://github.com/firebase/php-jwt/issues/543#issuecomment-1842592433

The spec seems to say items which do not have `expiry` set explicitly should be interpreted as never expiring (evne once retrieved from the cache). I don't know if I agree with this, but I think we should be consistent with the spec!

Reply:
https://github.com/symfony/symfony/issues/26241#issuecomment-366975853

PSR-6 is beautiful, innit?

Reply:
Updated!